### DISCUSSION TITLE
How to correct type tuple for refine args

### DISCUSSION THREAD

**Discussion by kricha** [colinhacks/zod#2944] [Q&A] [answered]
Hey, I've created reusable pass/confirm pass schema and refine, but can't find clear way for that. Or I just need help with type for tuple. Thank

```ts
const passwordWithConfirmZodObj = z.object({
  password: z.string().min(10, 'Password must be at least 10 characters'),
  confirmPassword: z.string(),
});

const confirmPasswordRefine: [Function, object] = [
  (data: z.infer<typeof passwordWithConfirmZodObj>) => data.password === data.confirmPassword,
  {
    message: 'Passwords must match',
    path: ['confirmPassword'],
  },
];

export const signUpSchema = z
  .object({
    email: z.string().email(),
    referral: z.string().nullable(),
  })
  .merge(passwordWithConfirmZodObj)
  .refine(...confirmPasswordRefine);

```

and on refine method I have error:

> No overload matches this call.
>   Overload 1 of 2, '(check: (arg: { password: string; confirmPassword: string; email: string; referral: string | null; }) => arg is { password: string; confirmPassword: string; email: string; referral: string | null; }, message?: string | ... 2 more ... | undefined): ZodEffects<...>', gave the following error.
>     Argument of type 'Function' is not assignable to parameter of type '(arg: { password: string; confirmPassword: string; email: string; referral: string | null; }) => arg is { password: string; confirmPassword: string; email: string; referral: string | null; }'.
>       Type 'Function' provides no match for the signature '(arg: { password: string; confirmPassword: string; email: string; referral: string | null; }): arg is { password: string; confirmPassword: string; email: string; referral: string | null; }'.
>   Overload 2 of 2, '(check: (arg: { password: string; confirmPassword: string; email: string; referral: string | null; }) => unknown, message?: string | Partial<Omit<ZodCustomIssue, "code">> | ((arg: { ...; }) => Partial<...>) | undefined): ZodEffects<...>', gave the following error.
>     Argument of type 'Function' is not assignable to parameter of type '(arg: { password: string; confirmPassword: string; email: string; referral: string | null; }) => unknown'.
>       Type 'Function' provides no match for the signature '(arg: { password: string; confirmPassword: string; email: string; referral: string | null; }): unknown'.ts(2769)

or if I give any type for tuple:

> A spread argument must either have a tuple type or be passed to a rest parameter.ts(2556)

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
function makePasswordWithConfirm () {
    const schema = z.object( {
        password: z.string().min( 10, 'Password must be at least 10 characters' ),
        confirmPassword: z.string(),
    } )
    const refineTuple: [ check: ( data: z.infer<typeof schema> ) => boolean, message: z.CustomErrorParams ] = [
        data => data.password === data.confirmPassword,
        { message: 'Passwords must match', path: [ 'confirmPassword' ] },
    ]
    return { schema, refineTuple }
}

const passwordWithConfirm = makePasswordWithConfirm()

const signUpSchema = z.object( {
    email: z.string().email(),
    referral: z.string().nullable(),
} )
    .merge( passwordWithConfirm.schema )
    .refine( ...passwordWithConfirm.refineTuple )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by kricha** [reply]
thanks!
